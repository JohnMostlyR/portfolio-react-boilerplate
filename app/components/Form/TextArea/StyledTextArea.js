import styled from 'styled-components';
import PropTypes from 'prop-types';

import { theme } from '../theme';
import { iconUrl } from '../HandIcon';
import { BASE_FONT_REGULAR } from '../typography';

export const StyledTextArea = styled.textarea`
  height: ${({ myHeight }) => (myHeight ? `${myHeight}px` : '2em')};
  margin: 0 !important;
  width: 100%;
  appearance: none;
  background-color: ${theme.backgroundColor};
  border: 0;
  border-bottom: 1px dotted;
  color: inherit;
  font-size: 1rem;
  outline: none;
  overflow: hidden;
  padding: 0.5rem 1.5rem 0.5rem 0;
  resize: none;
  ${BASE_FONT_REGULAR} @media (min-width: 600px) {
    background-image: ${() => `url(${iconUrl})`};
    background-position: calc(100% + 1.5em) 50%;
    background-repeat: no-repeat;
    background-size: 1.5em;
    transition: background-position 0.3s;

    &:focus {
      background-position: center right;
      border-bottom-style: solid;
    }
  }
`;

StyledTextArea.propTypes = {
  myHeight: PropTypes.number,
};

export default StyledTextArea;
