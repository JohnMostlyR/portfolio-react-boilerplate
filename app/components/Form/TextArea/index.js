/**
 *
 * TextArea
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import messages from '../messages';
import InputGroup from '../InputGroup';
import StyledTextArea from './StyledTextArea';
import Label from '../Label';
import Placeholder from '../Placeholder';
import LabelContent from '../LabelContent';
import ErrorMessage from '../ErrorMessage';
import SubText from '../SubText';

/* eslint-disable no-underscore-dangle */
class TextArea extends React.Component {
  constructor(props) {
    super(props);

    this.inputElement = React.createRef();
    this.state = {
      _error: false,
      _hasFocus: false,
      _height: 0,
      _value: this.props.value,
    };

    this.setFocus = this.setFocus.bind(this);
    this.handleFocusEvent = this.handleFocusEvent.bind(this);
    this.handleBlurEvent = this.handleBlurEvent.bind(this);
    this.handleChangeEvent = this.handleChangeEvent.bind(this);
  }

  componentWillReceiveProps(update) {
    this.setState({ _value: update.value });
  }

  setFocus() {
    if (this.inputElement) {
      this.inputElement.focus();
    }
  }

  handleFocusEvent() {
    this.setState({ _hasFocus: true });
  }

  handleBlurEvent() {
    const _hasFocus = false;

    this.setState({
      _hasFocus,
    });
  }

  handleChangeEvent(evt) {
    const { target } = evt;
    const { value } = target;
    const { name, validate } = this.props;
    const error = validate ? validate(value) : false;
    const { clientHeight, scrollHeight } = this.inputElement;
    let { _height } = this.state;

    // Adjust height of textarea on multi-line input
    if (scrollHeight > clientHeight) {
      _height = scrollHeight;
    } else {
      _height = clientHeight;
    }

    this.setState({
      _error: error,
      _height,
      _value: value,
    });

    this.props.changeHandler({
      name,
      value,
      error,
    });
  }

  render() {
    const {
      helperText,
      label,
      maxLength,
      minLength,
      name,
      placeholder,
    } = this.props;

    const { _error, _hasFocus, _height, _value } = this.state;

    const renderHelperText = () => {
      if (!helperText) return '';

      let helperTextRange = '';

      if (minLength || maxLength) {
        helperTextRange = (
          <FormattedMessage
            {...messages.range}
            values={{
              minLength,
              maxLength,
              count: _value.length,
            }}
          />
        );
      }

      if (_hasFocus || _error || !_value) {
        return (
          <span>
            {helperText}&nbsp;{helperTextRange}
          </span>
        );
      } else if (!_error) {
        return <FormattedMessage {...messages.valid} />;
      }

      return '';
    };

    const hasFocusOrValue = _hasFocus || !!_value;

    return (
      <InputGroup>
        <Label htmlFor="name" onClick={this.setFocus}>
          <LabelContent hasFocus={hasFocusOrValue}>
            {label}
            <Placeholder hasFocus={hasFocusOrValue}>
              <i>{placeholder}</i>
            </Placeholder>
          </LabelContent>
        </Label>
        <StyledTextArea
          myHeight={_height}
          name={name}
          onBlur={this.handleBlurEvent}
          onChange={this.handleChangeEvent}
          onFocus={this.handleFocusEvent}
          value={_value}
          innerRef={el => {
            this.inputElement = el;
          }}
        />
        <SubText>
          <ErrorMessage
            showError={!!(!_hasFocus && _error)}
          >{`${_error} `}</ErrorMessage>
          <span>{renderHelperText()}</span>
        </SubText>
      </InputGroup>
    );
  }
}

TextArea.propTypes = {
  helperText: PropTypes.string,
  label: PropTypes.string.isRequired,
  maxLength: PropTypes.number,
  minLength: PropTypes.number,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  changeHandler: PropTypes.func.isRequired,
  validate: PropTypes.func,
  value: PropTypes.string,
};

export default TextArea;
