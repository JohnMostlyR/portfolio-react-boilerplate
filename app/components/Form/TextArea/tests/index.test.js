import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import TextArea from '../index';
import StyledInput from '../StyledTextArea';

/* eslint no-underscore-dangle: "off" */
describe('<TextArea />', () => {
  const props = {
    helperText: 'This is a helper text',
    inputType: 'dummy',
    label: 'This is the label text',
    maxLength: 300,
    name: 'This is the input its name',
    minLength: 1,
    changeHandler: jest.fn(),
    placeholder: 'This is the placeholder text',
    validate: jest.fn(),
    value: 'The value',
  };

  it('should render with only the required properties and match the snapshot', () => {
    const wrapper = shallow(
      <TextArea
        name={props.name}
        label={props.label}
        changeHandler={props.changeHandler}
      />,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should receive all properties', () => {
    const wrapper = shallow(<TextArea {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should have the correct initial state', () => {
    const wrapper = shallow(<TextArea {...props} />);

    expect(wrapper.state()._error).toBe(false);
    expect(wrapper.state()._hasFocus).toBe(false);
    expect(wrapper.state()._height).toBe(0);
    expect(wrapper.state()._value).toBe(props.value);
  });

  it('should act when receiving focus', () => {
    const wrapper = shallow(<TextArea {...props} />);
    const input = wrapper.find(StyledInput);

    wrapper.setState({ _hasFocus: false });

    input.simulate('focus', { target: { value: true } });

    expect(wrapper.state()._hasFocus).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('when a user populates the field', () => {
    props.validate.mockReturnValue('Error, Error, Error');

    const wrapper = shallow(<TextArea {...props} />);
    const input = wrapper.find(StyledInput);
    const value = 'testing123\ntesting\ntesting\n123';

    beforeEach(() => {
      wrapper.setState({
        _error: false,
        _hasFocus: true,
        _height: 0,
        _value: '',
      });

      input.simulate('change', {
        target: {
          value,
        },
      });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should store the user input in the state property `_value`', () => {
      expect(wrapper.state()._value).toBe(value);
    });

    it('should update the helper text, when available', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should validate the user input', () => {
      expect(props.validate).toHaveBeenCalledTimes(1);
      expect(props.validate).toHaveBeenCalledWith(value);
    });

    it('should not show any errors as long as the field has the focus', () => {
      wrapper.setState({ _error: 'Error message.' });

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should call the `props.changeHandler` function', () => {
      expect(props.changeHandler).toHaveBeenCalledTimes(1);
      expect(props.changeHandler).toHaveBeenCalledWith({
        name: props.name,
        value,
        error: 'Error, Error, Error',
      });
    });

    it('should adjust the height when needed', () => {
      wrapper.setState({ _height: 32 });
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('when moving away', () => {
    const wrapper = shallow(<TextArea {...props} />);
    const input = wrapper.find(StyledInput);

    beforeEach(() => {
      wrapper.setState({
        _error: false,
        _hasFocus: true,
        _value: '',
      });
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should update state', () => {
      input.simulate('blur', {
        target: {
          value: true,
        },
      });

      expect(wrapper.state()._hasFocus).toBe(false);
    });

    it('should show an error message when the input is not correct', () => {
      wrapper.setState({
        _error: 'Error, Error, Error',
      });

      input.simulate('blur', {
        target: {
          value: '',
        },
      });

      expect(wrapper).toMatchSnapshot();
    });
  });
});
