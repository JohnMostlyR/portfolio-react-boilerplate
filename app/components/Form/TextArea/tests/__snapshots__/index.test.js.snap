// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextArea /> should act when receiving focus 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={true}
    >
      This is the label text
      <Placeholder
        hasFocus={true}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    myHeight={0}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="The value"
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      false 
    </ErrorMessage>
    <span>
      <span>
        This is a helper text
         
        <FormattedMessage
          defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
          id="portfolio.components.Form.helperText.range"
          values={
            Object {
              "count": 9,
              "maxLength": 300,
              "minLength": 1,
            }
          }
        />
      </span>
    </span>
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> should receive all properties 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={true}
    >
      This is the label text
      <Placeholder
        hasFocus={true}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    myHeight={0}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="The value"
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      false 
    </ErrorMessage>
    <span>
      <FormattedMessage
        defaultMessage="Thanks!"
        id="portfolio.components.Form.helperText.valid"
        values={Object {}}
      />
    </span>
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> should render with only the required properties and match the snapshot 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={false}
    >
      This is the label text
      <Placeholder
        hasFocus={false}
      >
        <i />
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    myHeight={0}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      false 
    </ErrorMessage>
    <span />
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> when a user populates the field should adjust the height when needed 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={true}
    >
      This is the label text
      <Placeholder
        hasFocus={true}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    myHeight={32}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="testing123
testing
testing
123"
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      Error, Error, Error 
    </ErrorMessage>
    <span>
      <span>
        This is a helper text
         
        <FormattedMessage
          defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
          id="portfolio.components.Form.helperText.range"
          values={
            Object {
              "count": 30,
              "maxLength": 300,
              "minLength": 1,
            }
          }
        />
      </span>
    </span>
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> when a user populates the field should not show any errors as long as the field has the focus 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={true}
    >
      This is the label text
      <Placeholder
        hasFocus={true}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="testing123
testing
testing
123"
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      Error message. 
    </ErrorMessage>
    <span>
      <span>
        This is a helper text
         
        <FormattedMessage
          defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
          id="portfolio.components.Form.helperText.range"
          values={
            Object {
              "count": 30,
              "maxLength": 300,
              "minLength": 1,
            }
          }
        />
      </span>
    </span>
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> when a user populates the field should update the helper text, when available 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={true}
    >
      This is the label text
      <Placeholder
        hasFocus={true}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value="testing123
testing
testing
123"
  />
  <SubText>
    <ErrorMessage
      showError={false}
    >
      Error, Error, Error 
    </ErrorMessage>
    <span>
      <span>
        This is a helper text
         
        <FormattedMessage
          defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
          id="portfolio.components.Form.helperText.range"
          values={
            Object {
              "count": 30,
              "maxLength": 300,
              "minLength": 1,
            }
          }
        />
      </span>
    </span>
  </SubText>
</InputGroup>
`;

exports[`<TextArea /> when moving away should show an error message when the input is not correct 1`] = `
<InputGroup>
  <Label
    htmlFor="name"
    onClick={[Function]}
  >
    <LabelContent
      hasFocus={false}
    >
      This is the label text
      <Placeholder
        hasFocus={false}
      >
        <i>
          This is the placeholder text
        </i>
      </Placeholder>
    </LabelContent>
  </Label>
  <StyledTextArea
    innerRef={[Function]}
    myHeight={0}
    name="This is the input its name"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    value=""
  />
  <SubText>
    <ErrorMessage
      showError={true}
    >
      Error, Error, Error 
    </ErrorMessage>
    <span>
      <span>
        This is a helper text
         
        <FormattedMessage
          defaultMessage="Should be between {minLength, number} and {maxLength, number} characters. Is now: {count, number}"
          id="portfolio.components.Form.helperText.range"
          values={
            Object {
              "count": 0,
              "maxLength": 300,
              "minLength": 1,
            }
          }
        />
      </span>
    </span>
  </SubText>
</InputGroup>
`;
